pkg/smi/traffic_target.go

```diff
@@ -0,0 +1,168 @@
+}
+
+// FixIt implements common.Runnable
+func (check TrafficTargetCheck) FixIt() error {
+	panic("implement me")
+}
+
+func doesTargetMatchPods(spec access.TrafficTargetSpec, srcPod *corev1.Pod, dstPod *corev1.Pod) bool {
+	// Map traffic targets to the given pods
+	if !cli.DoesTargetRefDstPod(spec, dstPod) {
+		return false
+	}
+	// The TrafficTarget destination is associated to 'dstPod', check if 'srcPod` is an allowed source to this destination
+	if cli.DoesTargetRefSrcPod(spec, srcPod) {
+		return true
+	}
+	return false
+}
+
+// Verify interface compliance
+var _ common.Runnable = (*RoutesValidityCheck)(nil)
+
+// RoutesValidityCheck implements common.Runnable
+type RoutesValidityCheck struct {
+	cfg          configurator.Configurator
+	srcPod       *corev1.Pod
+	dstPod       *corev1.Pod
+	accessClient smiAccessClient.Interface
+}
+
+// AreTrafficRoutesValid checks whether TrafficTargets matching the src and dest pods have supported routes
+func AreTrafficRoutesValid(osmConfigurator configurator.Configurator, srcPod *corev1.Pod, dstPod *corev1.Pod, smiAccessClient smiAccessClient.Interface) RoutesValidityCheck {
+	return RoutesValidityCheck{
+		cfg:          osmConfigurator,
+		srcPod:       srcPod,
+		dstPod:       dstPod,
+		accessClient: smiAccessClient,
+	}
+}
+
+const (
+	HTTPRouteGroupKind = "HTTPRouteGroup"
+	TCPRouteKind       = "TCPRoute"
+)
+
+// Description implements common.Runnable
+func (check RoutesValidityCheck) Description() string {
+	return fmt.Sprintf("Checking whether Traffic Targets with source pod %s and destination pod %s have valid routes (%s or %s)", check.srcPod.Name, check.dstPod.Name, HTTPRouteGroupKind, TCPRouteKind)
+}
+
+// Run implements common.Runnable
+func (check RoutesValidityCheck) Run() outcomes.Outcome {
+	// Check if permissive mode is enabled, in which case every meshed pod is allowed to communicate with each other
+	if check.cfg.IsPermissiveTrafficPolicyMode() {
+		return outcomes.DiagnosticOutcome{LongDiagnostics: "OSM is in permissive traffic policy modes -- all meshed pods can communicate and SMI access policies are not applicable"}
+	}
+	trafficTargets, err := check.accessClient.AccessV1alpha3().TrafficTargets(check.dstPod.Namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return outcomes.FailedOutcome{Error: err}
+	}
+	unsupportedRouteTargets := map[string]string{}
+	for _, trafficTarget := range trafficTargets.Items {
+		spec := trafficTarget.Spec
+		if !doesTargetMatchPods(spec, check.srcPod, check.dstPod) {
+			continue
+		}
+		for _, rule := range spec.Rules {
+			kind := rule.Kind
+			if !(kind == HTTPRouteGroupKind || kind == TCPRouteKind) {
+				unsupportedRouteTargets[trafficTarget.Name] = kind
+			}
+		}
+	}
+	if len(unsupportedRouteTargets) > 0 {
+		errorString := check.generateErrorMessage(unsupportedRouteTargets)
+		return outcomes.FailedOutcome{Error: fmt.Errorf(errorString)}
+	}
+	return outcomes.SuccessfulOutcomeWithoutDiagnostics{}
```

Comment:
At this point we will know that there is a `*Route` referenced, which is supported by OSM Controller.

We need to take the next step and check for the **existence** of said `*Route`

For instance - we find that some TrafficTarget, which references our source && destination pod; Next step is to see whether the HTTPRoute we find referenced in said TrafficTarget actually exists.  For that we could use the HTTPRoute List function.

Here is an example:  https://github.com/openservicemesh/osm/blob/c05c35aea5d10b11237975c901bfe6ed3a0f8eca/pkg/smi/client.go#L175